libraries/AP_ExternalAHRS/AP_ExternalAHRS_AdvancedNavigation.cpp

```diff
@@ -0,0 +1,1613 @@
+    AN_DEVICE_NAME_ID,
+    AN_END_CONFIGURATION_PACKETS
+};
+
+/*
+    Vehicle Type provided to Advanced Navigation EKF
+
+    https://docs.advancednavigation.com/certus/ANPP/FilterOptionsPacket.htm?Highlight=Vehicle%20Type#Vehicle_Types
+*/
+enum AN_VehicleType : uint8_t
+{
+    VEHICLE_TYPE_UNLIMITED,
+    VEHICLE_TYPE_BICYCLE,
+    VEHICLE_TYPE_CAR,
+    VEHICLE_TYPE_HOVERCRAFT,
+    VEHICLE_TYPE_SUBMARINE,
+    VEHICLE_TYPE_3D_UNDERWATER,
+    VEHICLE_TYPE_FIXED_WING_PLANE,
+    VEHICLE_TYPE_3D_AIRCRAFT,
+    VEHICLE_TYPE_HUMAN,
+    VEHICLE_TYPE_SMALL_BOAT,
+    VEHICLE_TYPE_SHIP,
+    VEHICLE_TYPE_STATIONARY,
+    VEHICLE_TYPE_STUNT_PLANE,
+    VEHICLE_TYPE_RACE_CAR
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// ANPP Packets
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+/*
+    ANPP Acknowledge Packet (0)
+
+    https://docs.advancednavigation.com/certus/ANPP/SystemPackets.htm#Acknowledge_Packet
+*/
+struct PACKED AN_Acknowledge
+{
+    enum AN_AcknowledgeResult : uint8_t
+    {
+        AN_ACKNOWLEDGE_SUCCESS = 0,
+        AN_ACKNOWLEDGE_CRC_ERROR = 1,
+        AN_ACKNOWLEDGE_PACKET_SIZE_ERROR = 2,
+        AN_ACKNOWLEDGE_RANGE_ERROR = 3,
+        AN_ACKNOWLEDGE_SYSTEM_FLASH_ERROR = 4,
+        AN_ACKNOWLEDGE_SYSTEM_NOT_READY = 5,
+        AN_ACKNOWLEDGE_UNKNOWN_PACKET = 6,
+    };
+
+    AN_PacketId id_acknowledged; // packet id being acknowledge
+    uint16_t crc_acknowledged;   // crc of packet being acknowledge
+    AN_AcknowledgeResult result; // acknowledgement result
+
+    const char *get_result_pretty() const
+    {
+        switch (result)
+        {
+        case AN_ACKNOWLEDGE_SUCCESS:
+            return "Success [0]";
+        case AN_ACKNOWLEDGE_CRC_ERROR:
+            return "CRC error [1]";
+        case AN_ACKNOWLEDGE_PACKET_SIZE_ERROR:
+            return "Packet size incorrect [2]";
+        case AN_ACKNOWLEDGE_RANGE_ERROR:
+            return "Values outside of valid range [3]";
+        case AN_ACKNOWLEDGE_SYSTEM_FLASH_ERROR:
+            return "System flash memory failure [4]";
+        case AN_ACKNOWLEDGE_SYSTEM_NOT_READY:
+            return "System not ready [5]";
+        case AN_ACKNOWLEDGE_UNKNOWN_PACKET:
+            return "Unknown packet [6]";
+        default:
+            return "Unknown error";
+        }
+    }
+};
+
+/*
+    ANPP Request Packet (1)
+
+    The request packet can contain multiple requested packets (up to the size of
+    the ANPP paylod (i.e. 255))
+
+    https://docs.advancednavigation.com/certus/ANPP/SystemPackets.htm#Request_Packet
+*/
+struct PACKED AN_RequestPacket
+{
+    AN_PacketId packet_id[AN_PacketPayload::AN_MAXIMUM_PACKET_SIZE];
+
+    uint8_t generate(uint8_t count, uint8_t id, ...)
+    {
+        va_list args;
+        va_start(args, id);
+
+        auto current_id = id;
+        for (unsigned i = 0; i < count; i++)
+        {
```

Comment:
```suggestion
        for (auto i = 0; i < count; i++) {
```

similar type and formatting changes everywhere, please